saml/src/main/java/com/linecorp/armeria/saml/SamlAssertionConsumerFunction.java

```diff
@@ -0,0 +1,265 @@
+        this.myEntityId = myEntityId;
+        this.idpConfigs = idpConfigs;
+        this.authManager = authManager;
+        this.sessionManager = sessionManager;
+        this.requestIdManager = requestIdManager;
+    }
+
+    @Override
+    public HttpResponse apply(ServiceRequestContext ctx, AggregatedHttpMessage msg) {
+        try {
+            // If this message is sent via HTTP-redirect binding protocol, its signature parameter should
+            // be validated.
+            final MessageContext<Response> messageContext;
+            if (cfg.endpoint().bindingProtocol().isHttpRedirect()) {
+                messageContext = HttpRedirectBindingUtil.toSamlObject(msg, PARAM_SAML_RESPONSE, idpConfigs);
+            } else {
+                messageContext = HttpPostBindingUtil.toSamlObject(msg, PARAM_SAML_RESPONSE);
+            }
+
+            final Response response = messageContext.getMessage();
+            final Assertion assertion = getValidatedAssertion(response);
+
+            // Find a session index which is sent by an identity provider.
+            final String sessionIndex = assertion.getAuthnStatements().stream()
+                                                 .map(AuthnStatement::getSessionIndex)
+                                                 .filter(Objects::nonNull)
+                                                 .findFirst().orElse(null);
+
+            final SAMLBindingContext bindingContext =
+                    messageContext.getSubcontext(SAMLBindingContext.class);
+            final String relayState = bindingContext != null ? bindingContext.getRelayState() : null;
+
+            return HttpResponse.from(sessionManager.signIn(messageContext, sessionIndex).handle(
+                    (unused, cause) -> cause != null ? HttpResponse.of(HttpStatus.SERVICE_UNAVAILABLE)
+                                                     : authManager.onSuccess(ctx, messageContext,
+                                                                             relayState)));
+        } catch (Exception e) {
+            logger.error("cannot consume an assertion", e);
+            return authManager.onFailure(ctx, e);
+        }
+    }
+
+    private Assertion getValidatedAssertion(Response response) throws SAMLException {
+        final Status status = response.getStatus();
+        if (!StatusCode.SUCCESS.equals(status.getStatusCode().getValue())) {
+            throw new SAMLException("status: " + status);
+        }
+
+        // The <Response> element MUST conform to the following:
+        //
+        // - The <Issuer> element MAY be omitted, but if present it MUST contain the unique identifier
+        //   of the issuing identity provider; the Format attribute MUST be omitted or have a value of
+        //   urn:oasis:names:tc:SAML:2.0:nameid-format:entity.
+        final Issuer issuer = response.getIssuer();
+        if (issuer == null || issuer.getValue() == null) {
+            throw new SAMLException("no issuer");
```

Comment:
I may not be understanding the comment above, but is it ok to throw an exception even though `<Issuer> element MAY be omitted`?

Reply:
It was my fault. Thank you!